cache_archive.go

```diff
@@ -0,0 +1,252 @@
+		fmt.Println()
+		log.Warnf("First upload attempt failed, retrying...")
+		fmt.Println()
+		time.Sleep(3000 * time.Millisecond)
+		return tryToUploadArchive(uploadURL, pth)
+	}
+	return nil
+}
+
+// getCacheUploadURL requests an upload url from the Bitrise cache API server.
+func getCacheUploadURL(cacheAPIURL string, fileSizeInBytes int64) (string, error) {
+	if strings.HasPrefix(cacheAPIURL, "file://") {
+		return cacheAPIURL, nil
+	}
+
+	req, err := http.NewRequest("POST", cacheAPIURL, bytes.NewReader([]byte(fmt.Sprintf(`{"file_size_in_bytes": %d}`, fileSizeInBytes))))
+	if err != nil {
+		return "", fmt.Errorf("failed to create request: %s", err)
+	}
+
+	resp, err := (&http.Client{Timeout: 20 * time.Second}).Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to send request: %s", err)
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			log.Warnf("Failed to close response body: %s", err)
+		}
+	}()
+
+	if resp.StatusCode < 200 || resp.StatusCode > 202 {
+		return "", fmt.Errorf("upload url was rejected with status code: %d", resp.StatusCode)
+	}
+
+	var respModel map[string]string
+	if err := json.NewDecoder(resp.Body).Decode(&respModel); err != nil {
+		return "", fmt.Errorf("failed to decode response body: %s", err)
+	}
+
+	uploadURL, ok := respModel["upload_url"]
+	if !ok || uploadURL == "" {
+		return "", fmt.Errorf("request sent, but upload url isn't received")
+	}
+
+	return uploadURL, nil
+}
+
+// tryToUploadArchive performs the cache upload.
+// If the destination is a local file path (url has a file:// scheme) this function copies the cache archive file to the destination.
+// Otherwise destination should be a remote url.
+func tryToUploadArchive(uploadURL string, archiveFilePath string) error {
+	if strings.HasPrefix(uploadURL, "file://") {
+		pth := strings.TrimPrefix(uploadURL, "file://")
+		return command.CopyFile(archiveFilePath, pth)
+	}
+
+	archFile, err := os.Open(archiveFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to open archive file for upload (%s): %s", archiveFilePath, err)
+	}
+
+	fileClosed := false
+	defer func() {
+		if fileClosed {
+			return
+		}
+		if err := archFile.Close(); err != nil {
+			log.Warnf("Failed to close archive file (%s): %s", archiveFilePath, err)
+		}
+	}()
+
+	fileInfo, err := archFile.Stat()
+	if err != nil {
+		return fmt.Errorf("failed to get file stats of the archive file (%s): %s", archiveFilePath, err)
+	}
+	fileSize := fileInfo.Size()
+
+	req, err := http.NewRequest("PUT", uploadURL, archFile)
+	if err != nil {
+		return fmt.Errorf("failed to create upload request: %s", err)
+	}
+
+	req.Header.Add("Content-Length", strconv.FormatInt(fileSize, 10))
+	req.ContentLength = fileSize
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to upload: %s", err)
+	}
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("upload failed with status code: %d", resp.StatusCode)
+	}
+
+	fileClosed = true
```

Comment:
Where, how is `archFile` closed? The defer statement above (https://github.com/bitrise-steplib/steps-cache-push/pull/21/files#diff-b73d81dbfb105e13e6baeaec05b57b4fR218) would always skip closing it, because of this assignment.

Reply:
It will be closed here: https://github.com/bitrise-steplib/steps-cache-push/pull/21/files/2c05f01542ed2cb912f7c55bfb754682fe98ea46#diff-b73d81dbfb105e13e6baeaec05b57b4fR240
by the `http.DefaultClient.Do(req)`

[NewRequest](https://golang.org/pkg/net/http/#NewRequest)

_If the provided body is also an io.Closer, the returned Request.Body is set to body and will be closed by the Client methods Do, Post, and PostForm, and Transport.RoundTrip._